### PR TITLE
Real Time TA

### PR DESCRIPTION
**Pull request by tushdante**
- Added support for the Real Time Audiences endpoint
- Currently only works with `.csv` files containing hashed IDs

Current Process:
- Read the file
- Create a list of JSON objects with each object being a single membership update/delete
- Chunk the list into batches of 100 (the max batch size)
- Make API calls

See the example for more details

### CODE REVIEW

File: twitter_ads/resource.py [hunk 1]
```diff
@@ -177,6 +178,48 @@ def batch_save(klass, account, objs):
...
+        return response.body['data'][0]['success_count']
+
+    @classmethod
+    def ta_batch(klass, file_path, params):
+        """
+        Returns an iterator TA member entites for a passed in csv file
+        """
+        ta_member_list = []
+        with open(file_path, 'r') as file:
+            reader = csv.reader(file, dialect='excel')
```
Comment (jbabichjapan) [inline twitter_ads/resource.py]: Can you confirm this is definitely the dialect we want to use

File: twitter_ads/audience.py [hunk 2]
```diff
@@ -35,6 +37,44 @@ def create(klass, account, file_path, name, list_type):
             raise e
 
     @classmethod
+    def ta_file_upload(klass, account, file_path, name, list_type, expires_at, effective_at):
```
Comment (jbabichjapan) [inline twitter_ads/audience.py]: We might as well include a sample file with dummy data if we are going to include a file based method - i.e. it's confusing whether or not the data in the file should be pre-hashed and if it's pre-hashed then maybe they need a utility to do that without mangling the other parts of the file (commas, etc)

File: twitter_ads/audience.py [hunk 3]
```diff
@@ -35,6 +37,44 @@ def create(klass, account, file_path, name, list_type):
...
+        ta_members = klass.ta_batch(file_path, params)
+
+        # create groups of 100 entities per batch
+        ta_member_chunked = klass.chunkify(ta_members, klass.MAX_CHUNK_SIZE)
+
+        # run through the iterator and upload each of the objects
+        success_count = 0
+
+        for member in ta_member_chunked:
+            success_count += klass.ta_batch_save(account, member)
```
Comment (jbabichjapan) [inline twitter_ads/audience.py]: Is retry being handled correctly?

**Comment by juanshishido** [comment]
Will review this soon, but wanted to note that the [Ruby SDK #144](https://github.com/twitterdev/twitter-ruby-ads-sdk/pull/144) has had, from what I can tell, similar functionality added.

File: twitter_ads/resource.py [hunk 4]
```diff
@@ -177,6 +178,48 @@ def batch_save(klass, account, objs):
...
+            for text in reader:
+                for line in text:
+                    if not line:
```
Comment (juanshishido) [inline twitter_ads/resource.py]: Once it reaches the EOF, it'll just stop. No need to add the `break`.

File: twitter_ads/audience.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tushdante) [reply twitter_ads/audience.py]: Thats a good point. My initial reason for keeping this under the same class was in the event we'd want to migrate everyone over to the real-time endpoint. That being said, I think I prefer the idea of having a separate class instead.

**Comment by smaeda-ks** [comment]
Closeable? Real-Time TA will be sunset soon.

**Comment by smaeda-ks** [comment]
Closing this PR as it's already retired.

**Pull request closed without merging**